### PR TITLE
k8s: use kubeconfig instead of namespace in yml files fixes #9777

### LINKED ISSUE #9777: kubernetes namespace should not be present the yml files
##### **Overview of the issue**
Kubectl can't deploy to the right namespace since the valu ein the context is overridden by the yml file

##### **Motivation for or Use Case**
Having different namespaces for test and prod

the namespace can be defined in kubeconf per context so that each one uses it's own namespace
##### **Reproduce the error**
generate kubernetes

##### **Suggest a Fix**
remove kubernetesNamespace from the generator (or have it optional is you want)

##### **JHipster Version(s)**

5.0.1 this not a regression
##### **Browsers and Operating System**

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by yelhouti**
fix https://github.com/jhipster/generator-jhipster/issues/9777

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [X] Tests are added where necessary
-   [X] Documentation is added/updated where necessary
-   [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/kubernetes/templates/registry/consul-config-loader.yml.ejs [hunk 1]
```diff
@@ -39,7 +38,7 @@ spec:
         - name: INIT_SLEEP_SECONDS
           value: "5"
         - name: CONSUL_URL
-          value: consul.<%= kubernetesNamespace %>.svc.cluster.local
+          value: consul.
```
Comment (DanielFran) [inline generators/kubernetes/templates/registry/consul-config-loader.yml.ejs]: substitute "consul." by "consul"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 2]
```diff
@@ -72,14 +71,14 @@ spec:
           value: prod
         <%_ if (app.serviceDiscoveryType === 'eureka') { _%>
         - name: SPRING_CLOUD_CONFIG_URI
-          value: http://admin:${jhipster.registry.password}@jhipster-registry.<%= kubernetesNamespace %>.svc.cluster.local:8761/config
+          value: http://admin:${jhipster.registry.password}@jhipster-registry.:8761/config
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "http://admin:${jhipster.registry.password}@jhipster-registry.:8761/config" by "http://admin:${jhipster.registry.password}@jhipster-registry:8761/config"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 3]
```diff
@@ -72,14 +71,14 @@ spec:
...
         - name: JHIPSTER_REGISTRY_PASSWORD
           valueFrom:
             secretKeyRef:
               name: registry-secret
               key: registry-admin-password
         - name: EUREKA_CLIENT_SERVICE_URL_DEFAULTZONE
-          value: http://admin:${jhipster.registry.password}@jhipster-registry.<%= kubernetesNamespace %>.svc.cluster.local:8761/eureka/
+          value: http://admin:${jhipster.registry.password}@jhipster-registry.:8761/eureka/
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "http://admin:${jhipster.registry.password}@jhipster-registry.:8761/eureka/" by "http://admin:${jhipster.registry.password}@jhipster-registry:8761/eureka/"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 4]
```diff
@@ -89,7 +88,7 @@ spec:
         <%_ } _%>
         <%_ if (app.serviceDiscoveryType === 'consul') { _%>
         - name: SPRING_CLOUD_CONSUL_HOST
-          value: consul.<%= kubernetesNamespace %>.svc.cluster.local
+          value: consul.
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "consul." by "consul"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 5]
```diff
@@ -110,11 +109,11 @@ spec:
         <%_ } _%>
         <%_ if (app.prodDatabaseType === 'mysql') { _%>
         - name: SPRING_DATASOURCE_URL
-          value: jdbc:mysql://<%= app.baseName.toLowerCase() %>-mysql.<%= kubernetesNamespace %>.svc.cluster.local:3306/<%= app.baseName.toLowerCase() %>?useUnicode=true&characterEncoding=utf8&useSSL=false
+          value: jdbc:mysql://<%= app.baseName.toLowerCase() %>-mysql.:3306/<%= app.baseName.toLowerCase() %>?useUnicode=true&characterEncoding=utf8&useSSL=false
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "jdbc:mysql://<%= app.baseName.toLowerCase() %>-mysql.:3306/<%= app.baseName.toLowerCase() %>?useUnicode=true&characterEncoding=utf8&useSSL=false" by "jdbc:mysql://<%= app.baseName.toLowerCase() %>-mysql:3306/<%= app.baseName.toLowerCase() %>?useUnicode=true&characterEncoding=utf8&useSSL=false"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 6]
```diff
@@ -110,11 +109,11 @@ spec:
...
         <%_ } _%>
         <%_ if (app.prodDatabaseType === 'mariadb') { _%>
         - name: SPRING_DATASOURCE_URL
-          value: jdbc:mariadb://<%= app.baseName.toLowerCase() %>-mariadb.<%= kubernetesNamespace %>.svc.cluster.local:3306/<%= app.baseName.toLowerCase() %>
+          value: jdbc:mariadb://<%= app.baseName.toLowerCase() %>-mariadb.:3306/<%= app.baseName.toLowerCase() %>
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "jdbc:mariadb://<%= app.baseName.toLowerCase() %>-mariadb.:3306/<%= app.baseName.toLowerCase() %>" by "jdbc:mariadb://<%= app.baseName.toLowerCase() %>-mariadb:3306/<%= app.baseName.toLowerCase() %>"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 7]
```diff
@@ -123,7 +122,7 @@ spec:
         <%_ } _%>
         <%_ if (app.prodDatabaseType === 'postgresql') { _%>
         - name: SPRING_DATASOURCE_URL
-          value: jdbc:postgresql://<%= app.baseName.toLowerCase() %>-postgresql.<%= kubernetesNamespace %>.svc.cluster.local:5432/<%= app.baseName %>
+          value: jdbc:postgresql://<%= app.baseName.toLowerCase() %>-postgresql.:5432/<%= app.baseName %>
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "jdbc:postgresql://<%= app.baseName.toLowerCase() %>-postgresql.:5432/<%= app.baseName %>" by "jdbc:postgresql://<%= app.baseName.toLowerCase() %>-postgresql:5432/<%= app.baseName %>"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 8]
```diff
@@ -146,23 +145,23 @@ spec:
         <%_ } _%>
         <%_ if (app.searchEngine === 'elasticsearch') { _%>
         - name: SPRING_DATA_JEST_URI
-          value: http://<%= app.baseName.toLowerCase() %>-elasticsearch.<%= kubernetesNamespace %>.svc.cluster.local:9200
+          value: http://<%= app.baseName.toLowerCase() %>-elasticsearch.:9200
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "http://<%= app.baseName.toLowerCase() %>-elasticsearch.:9200" by "http://<%= app.baseName.toLowerCase() %>-elasticsearch:9200"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 9]
```diff
@@ -146,23 +145,23 @@ spec:
...
         - name: SPRING_ELASTICSEARCH_REST_URIS
-          value: http://<%= app.baseName.toLowerCase() %>-elasticsearch.<%= kubernetesNamespace %>.svc.cluster.local:9200
+          value: http://<%= app.baseName.toLowerCase() %>-elasticsearch.:9200
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "http://<%= app.baseName.toLowerCase() %>-elasticsearch.:9200" by "http://<%= app.baseName.toLowerCase() %>-elasticsearch:9200"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 10]
```diff
@@ -146,23 +145,23 @@ spec:
...
         <%_ } _%>
         <%_ if (app.messageBroker === 'kafka') { _%>
         - name: KAFKA_CONSUMER_KEY_DESERIALIZER
           value: 'org.apache.kafka.common.serialization.StringDeserializer'
         - name: KAFKA_CONSUMER_VALUE_DESERIALIZER
           value: 'org.apache.kafka.common.serialization.StringDeserializer'
         - name: KAFKA_CONSUMER_BOOTSTRAP_SERVERS
-          value: 'jhipster-kafka.<%= kubernetesNamespace %>.svc.cluster.local:9092'
+          value: 'jhipster-kafka.:9092'
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: Substitute "'jhipster-kafka.:9092'" by "'jhipster-kafka:9092'"

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 11]
```diff
@@ -146,23 +145,23 @@ spec:
...
         - name: KAFKA_CONSUMER_GROUP_ID
           value: '<%= app.baseName.toLowerCase() %>'
         - name: KAFKA_CONSUMER_AUTO_OFFSET_RESET
           value: 'earliest'
         - name: KAFKA_PRODUCER_BOOTSTRAP_SERVERS
-          value: 'jhipster-kafka.<%= kubernetesNamespace %>.svc.cluster.local:9092'
+          value: 'jhipster-kafka.:9092'
```
Comment (DanielFran) [inline generators/kubernetes/templates/deployment.yml.ejs]: SUbstitute "'jhipster-kafka.:9092'" by "'jhipster-kafka:9092'"

File: generators/kubernetes/templates/messagebroker/kafka.yml.ejs [hunk 12]
```diff
@@ -40,9 +39,9 @@ spec:
         image: <%= DOCKER_KAFKA %>
         env:
         - name: KAFKA_ADVERTISED_LISTENERS
-          value: 'PLAINTEXT://jhipster-kafka.<%= kubernetesNamespace %>.svc.cluster.local:9092'
+          value: 'PLAINTEXT://jhipster-kafka.:9092'
```
Comment (DanielFran) [inline generators/kubernetes/templates/messagebroker/kafka.yml.ejs]: Substitute "'PLAINTEXT://jhipster-kafka.:9092'" by "'PLAINTEXT://jhipster-kafka:9092'"

File: generators/kubernetes/templates/messagebroker/kafka.yml.ejs [hunk 13]
```diff
@@ -40,9 +39,9 @@ spec:
...
         - name: KAFKA_ZOOKEEPER_CONNECT
-          value: 'jhipster-zookeeper.<%= kubernetesNamespace %>.svc.cluster.local:2181'
+          value: 'jhipster-zookeeper.:2181'
```
Comment (DanielFran) [inline generators/kubernetes/templates/messagebroker/kafka.yml.ejs]: Substitute "'jhipster-zookeeper.:2181'" by "'jhipster-zookeeper:2181'"

File: generators/kubernetes/templates/registry/application-configmap.yml.ejs [hunk 14]
```diff
@@ -51,5 +50,5 @@ data:
     eureka:
       client:
         service-url:
-          defaultZone: http://admin:${spring.security.user.password}@jhipster-registry-0.jhipster-registry.<%= kubernetesNamespace %>.svc.cluster.local:8761/eureka/,http://admin:${spring.security.user.password}@jhipster-registry-1.jhipster-registry.<%= kubernetesNamespace %>.svc.cluster.local:8761/eureka/
+          defaultZone: http://admin:${spring.security.user.password}@jhipster-registry-0.jhipster-registry.:8761/eureka/,http://admin:${spring.security.user.password}@jhipster-registry-1.jhipster-registry.:8761/eureka/
```
Comment (DanielFran) [inline generators/kubernetes/templates/registry/application-configmap.yml.ejs]: Substitute "defaultZone: http://admin:${spring.security.user.password}@jhipster-registry-0.jhipster-registry.:8761/eureka/,http://admin:${spring.security.user.password}@jhipster-registry-1.jhipster-registry.:8761/eureka/" by "defaultZone: http://admin:${spring.security.user.password}@jhipster-registry-0.jhipster-registry:8761/eureka/,http://admin:${spring.security.user.password}@jhipster-registry-1.jhipster-registry:8761/eureka/"

**Comment by yelhouti** [comment]
@DanielFran Oups, thank for catching that, one comment is enough :)

**Comment by yelhouti** [comment]
@DanielFran Done :)

**Comment by PierreBesson** [comment]
I don't understand why we generate a kubeconfig file. Normally this file depends on your Kubernetes cluster. I don't really understand what you are trying to do but I think it cannot fit all use cases.

**Comment by yelhouti** [comment]
@PierreBesson I generate it for the same reason we generate all files, speed up the development process, and enforce best practices.

This file is written in a away to teach new users of kubernetes how to switch context..., you can still modify it with specific content..., also I hesitate to add a .gitignore file with ssl folder, to avoid users committing those..., but since they are not required (for other auth methods) I didn't

**Comment by PierreBesson** [comment]
@yelhouti but if it doesn't work out the box then there is no need to generate this file. It will only confuse the users. It's better to write this information in the docs.

File: generators/kubernetes/templates/kubeconfig.ejs [hunk 15]
```diff
@@ -0,0 +1,31 @@
+kind: Config
```
Comment (PierreBesson) [inline generators/kubernetes/templates/kubeconfig.ejs]: Please remove this file.

Comment (yelhouti) [reply generators/kubernetes/templates/kubeconfig.ejs]: sure, putting it in the README

Comment (PierreBesson) [reply generators/kubernetes/templates/kubeconfig.ejs]: Thanks

Comment (yelhouti) [reply generators/kubernetes/templates/kubeconfig.ejs]: Done

**Comment by deepu105** [comment]
I also don't like this approach. I do not want to meddle with my kubeconfig as well, Also having the namespace on the YAML was a best practice AFAIK, what is the advantage of this over having the NS?

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 16]
```diff
@@ -72,14 +71,14 @@ spec:
           value: prod
         <%_ if (app.serviceDiscoveryType === 'eureka') { _%>
         - name: SPRING_CLOUD_CONFIG_URI
-          value: http://admin:${jhipster.registry.password}@jhipster-registry.<%= kubernetesNamespace %>.svc.cluster.local:8761/config
+          value: http://admin:${jhipster.registry.password}@jhipster-registry:8761/config
```
Comment (deepu105) [inline generators/kubernetes/templates/deployment.yml.ejs]: would this still work if I specify a namespace during deployment?

**Comment by deepu105** [comment]
@srinivasa-vasu what do you think?

File: generators/kubernetes/templates/deployment.yml.ejs (hunk 16 above)
Comment (yelhouti) [reply generators/kubernetes/templates/deployment.yml.ejs]: of course

**Comment by yelhouti** [comment]
> Does everything go to default namespace now? 
Everything goes to what is specified in your kubeconfig (the file I was requested to remove, because everyone would know hoe to use one according to @PierreBesson ). I could put it back if you guys agree.
> Would this work if I set a default NS in config or pass it via `--namespace` flag during `kubectl apply`?
Yes it would

**Comment by deepu105** [comment]
We shouldn't generate any kubeconfig file. What we could do is add the namespace in the commands of the generated `kubectl-apply.sh` file, so the namespace user chooses during generation, is used when applying using the file. Anyways I have mixed feelings about this change. Let's see what @srinivasa-vasu @saturnism etc have to say

**Comment by deepu105** [comment]
@yelhouti did you test this with Istio, coz it also uses namepace in the generated files

**Comment by yelhouti** [comment]
@deepu105 I disagree, if we add them there, they would take presedence over kubeconfig,

**Comment by deepu105** [comment]
@yelhouti in that case you are kind of making the namespace question useless. IMO this change only makes it more confusing especially for novice users and I still don't see how this is more beneficial over having the namespace in the files

**Comment by yelhouti** [comment]
@deepu105 I checked the code and it seemed to me that istio uses it own namespace (in it's files), I also looked at all places that had cluster.local and removed it

**Comment by deepu105** [comment]
@yelhouti I meant if you look at Istio related files it uses the application namespaces in many places, especially in route URLs and stuff, for example, check `destination-rule.yml`, `gateway.yml` and so on, basically search for `<%= kubernetesNamespace %>` in the codebase

**Comment by yelhouti** [comment]
@deepu105 it was removed from there two, my point is as follow:

everything that should be deployed to the namespace chosen by the user, should not have the namespace in it, the namespace should be defined in the kubeconfig.

 I didn't invent this rule it is suggested by kubernetes team:

https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-organizing-with-namespaces

quotes:
```
If you specify a namespace in the YAML declaration, the resource will always be created in that namespace. If you try to use the “namespace” flag to set another namespace, the command will fail.
```

**Comment by yelhouti** [comment]
@deepu105 I use kubernets every week, and I'm pretty sure this is the right way to go...

**Comment by deepu105** [comment]
@yelhouti I don't see the k8s team suggesting this as a best practice, they are just mentioning that declaring NS in file means you cant set it via a flag, do they anywhere say that one practice is better than other? What they say is

> There are two ways to explicitly tell Kubernetes in which Namespace you want to create your resources.

> One way is to set the “namespace” flag when creating the resource:
> You can also specify a Namespace in the YAML declaration.
> If you specify a namespace in the YAML declaration, the resource will always be created in that namespace. If you try to use the “namespace” flag to set another namespace, the command will fail.

Anyway let's see what others think

**Comment by yelhouti** [comment]
@deepu105 sure, but if you think about it, the way it is said implies that, specifying it in the file makes it impossible to use the same config in two different namespace (ex: myProject-dev, myProject-prod). when not specifies, the kubeconfig allows to exactly that.

Check my comment in the k8s README: https://github.com/jhipster/generator-jhipster/pull/10545/files#diff-0

File: generators/kubernetes/templates/registry/application-configmap.yml.ejs (hunk 14 above)
Comment (yelhouti) [reply generators/kubernetes/templates/registry/application-configmap.yml.ejs]: All resolved, please remove change request

**Comment by srinivasa-vasu** [comment]
@yelhouti - I understand the reason behind removing the namespace tag. Both the approaches have their own pros and cons. The disadvantage of having namespace tag in the manifests is mitigated by the code generator tool. It so happens that if I want to apply the manifests to a specific namespace, then it needs to be passed as part of your kubectl command. But in my kubeconfig, if I set some other space as the default namespace and if I forget to pass the namespace in the command, then everything would get deployed to a namespace that I don't intend to do. That's the reason we have kept the namespace in all the files and it makes sense to do so while you are generating the code. This operation is idempotent, so if you need to generate it for a different space, just run the jhipster generator again. In that way, generated files control everything and I don't need to play around with the namespace tags. Hope this makes sense.

Really appreciate your thought process here, but personally I would prefer to keep the namespace tag as such.

**Comment by yelhouti** [comment]
@srinivasa-vasu , I undertsand what you propose, but it clearly goes against the code duplication principal.

If I have 3 environment (dev, staging, prod) I would need 3 projects, and any change in one of them must be re applied to the others, so your disadvantage is not really solved.

About the kubeconfig approach, the risk of using the wrong namespace disappears if you use sane defaults. (ie: never override you globale kubeconfig) and use one kubeconfig par project (or even namespace)

If your environement are different from each other and you want a script for running each one, all you need is:
```
set -e
export ./kubeconfig
kubectl config use-context dev
kubectl apply -f ...
kubectl apply -f ...
kubectl apply -f ...
kubectl config use-context default
```

Or if you are paranoid, you can even have a kubeconfig file per environment...
 
By the way, I used to do exactly what do until a more senior person convinced me of the opposite, and please don't get me wrong, I don't mind being hard to convince, alsmost all of use (in this field) are pretty hard to convince :)

**Comment by yelhouti** [comment]
@deepu105 Not necessarely, we could still use it to generate the kubeconfig. plus the place where it is still used are: 
- ingress, I think we could rename that variable,
- helm I think we should use the same approach
bur will still first need to agree on if ou guys want to switch. What do you think

**Comment by srinivasa-vasu** [comment]
@yelhouti - Don't you keep the namespace (name) same between dev, stage and prod? If so, then there is no duplication. But the reality is, that's always not the case, right? You might have to maintain config info per environment if you decide to use K8s manifests. If that's the case, you anyway have to maintain separate config files per environment.

**Comment by yelhouti** [comment]
@srinivasa-vasu in case you are a kubeconfig, then yes, it's one file per env, in case you hard code namespace into the file, you would need w whole different project.

**Comment by yelhouti** [comment]
@deepu105 @DanielFran  sorry for the ping, could you guys validate the changes each of you requested, or direct me on how I should proceed? thanks

File: generators/kubernetes/templates/istio/gateway.yml.ejs [hunk 17]
```diff
@@ -20,7 +20,6 @@ apiVersion: <%= KUBERNETES_ISTIO_NETWORKING_API_VERSION %>
 kind: Gateway
 metadata:
   name: <%= app.baseName.toLowerCase() %>-gateway
-  namespace: <%= kubernetesNamespace %>
```
Comment (deepu105) [inline generators/kubernetes/templates/istio/gateway.yml.ejs]: There is still `<%= app.baseName.toLowerCase() %>.<%= kubernetesNamespace %>.<%= ingressDomain %>` could you please check for all occurance of `kubernetesNamespace` in templates

**Comment by yelhouti** [comment]
@deepu105 thank for the feedback, I will rename the variable when needed if we decide to go with it.
I think the main concern here it to make sure that nothing is deployed to the wrong namespace, this is why we keep the config that an env variable that is reset in each terminal.
I'm not just trying to get the PR passed, I am really looking for your constructive feedback so please indulge me and my questions.
1- I know that kubectl comes with a default kubeconfig poiting to localhost, that you should not modify, so if you don't like kubeconfig en variable you can use `--kubeconfig=config-file`, so unless I am missing something, I don't see how you can work without a kubeconfig file to point to your cluster with your credentials ... isn't a custom kubeconfig kind of required wven for beginners?
2- kubernetes states for beginners that they should use the default namespace so I don't feel too bas about that: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#when-to-use-multiple-namespaces
> Namespaces are intended for use in environments with many users spread across multiple teams, or projects. For clusters with a few to tens of users, you should not need to create or think about namespaces at all. Start using namespaces when you need the features they provide.

File: generators/kubernetes/templates/istio/gateway.yml.ejs (hunk 17 above)
Comment (yelhouti) [reply generators/kubernetes/templates/istio/gateway.yml.ejs]: I will do as soon as we decide with @PierreBesson if we follow this route

File: generators/kubernetes/templates/README-KUBERNETES.md.ejs [hunk 18]
```diff
@@ -14,13 +14,53 @@ $ <%= dockerPushCommand %> <%= appConfigs[i].targetImageName %>
 ```
 
 ## Deployment
-
+Add the ssh keys for the user
 You can deploy all your apps by running the below bash command:
 
 ```
+export KUBECONFIG=./kubeconfig
```
[the remark this answers is not in the document]
Comment (yelhouti) [reply generators/kubernetes/templates/README-KUBERNETES.md.ejs]: I don't see any workflow where you could avoid using a kubeconfig file, you could have many and export one of them to switch between clusters/namespaces... (or pass it as a command argument) so sorry I feel like I don't totally understand your concern

**Comment by deepu105** [comment]
@yelhouti no worries, we are here to learn from each other.

Well, I use Kubernetes at my company a lot and our workflows are quite different. I(can say the same for my collegues) have honestly never used a kubeconfig file outside of the default(Execept for in automated tests) that is present in `~/.kube/config`. All my cluster information is stored in that file and I choose the `current-context` when I switch between them. Normally we never used same cluster to deploy different environment as our workloads were always on GKE, EKS or OpenShift, we have one cluster per environment and hence same namespace on all.


When I create a new cluster for example with GKE, I would execute `gcloud container clusters get-credentials <cluster-name>` or equivalent provided by EKS/AKS and so on so that the cluster details are added to my `~/.kube/config`. 

Also, I agree that NS is not a big deal unless you deploy multiple apps to the same cluster.

**Comment by yelhouti** [comment]
@deepu105 Oh I see thanks, so with my approach the different would be that your `~/.kube/config` would need to be modified to add one line: 
```
- name: context1
  context:
    cluster: mycluster
    namespace: mynamespace
    user: admin
```
Or if you use the same cluster for two things, you would need to copy the context 
- name: context-ns1
  context:
    cluster: mycluster
    namespace: ns1
    user: admin
- name: context-ns2
  context:
    cluster: mycluster
    namespace: ns2
    user: admin
```
And It should all work as it used to, am I right?

**Comment by deepu105** [comment]
Well, the thing is I don't want to add the namespace there, I rarely modify that file. But the real point is there is no universal workflow here, everyone follows a workflow that they are used to, we should just try to make what we generate as generic as possible without assuming any specific workflow and without the end-user having to fiddle around with any config just to get it working

**Comment by srinivasa-vasu** [comment]
I know there is a separate issue opened to remove the namespace tag. Though I am inclined towards removing the `service` local references, but really not from the templates. As we are generating the code, to me it makes sense to be self-contained. However, we can provide kustomize template on top to externalize things. Have a look at [this](https://github.com/jhipster/generator-jhipster/pull/11038)
@yelhouti @deepu105

**Comment by deepu105** [comment]
Looks good. I have not tried kustomize yet, will play around with it soon
and get back to you

**Comment by saturnism** [comment]
eak i def missed this conversation for a while.

my preferences had always been - no ns in the yaml file. if needed, specify ns during deployment. use context to deploy to different clusters.

and +1 to use kustomize for to change configuration and/or override namespaces for different environments.

**Comment by pascalgrimaud** [comment]
@deepu105 @PierreBesson @xetys @srinivasa-vasu @saturnism : what is the state of this PR ?
Do you all agree with this PR ? Should we merge this or not ?

As https://github.com/jhipster/generator-jhipster/issues/9777 is opened for too long, I'm trying to clean old tickets.

**Comment by pascalgrimaud** [comment]
@yelhouti : sorry for taking so long, @PierreBesson said me he'll have a last look on this.
Can you resolve the merge conflict plz ?

**Comment by yelhouti** [comment]
@pascalgrimaud when do you need it?

**Comment by pascalgrimaud** [comment]
@yelhouti : I plan to do the release this monday, unless someone in the core team is against, so I'd like to have this, if @PierreBesson agrees :)

**Comment by pascalgrimaud** [comment]
Sorry for this @yelhouti 
This PR is opened for too long (8 months), still conflicts, and no core team member want to merge this.
So that's why I'm closing this PR and the ticket too.

cc @jhipster/developers

**Pull request closed without merging**